stackslib/src/chainstate/stacks/boot/pox-4.clar

```diff
@@ -1311,6 +1323,62 @@
       (ok { stacker: stacker,
             unlock-burn-height: new-unlock-ht }))))
 
+;; This function allows a node to register a new 'current-signer-set' during the prepare phase
+(define-private (update-current-signer-set (new-signer-set (list 4096 {signer: (buff 33)})))
+    (let 
+        (
+            (next-reward-cycle (+ u1 current-pox-reward-cycle))
+            (next-reward-cycle-burn-height (reward-cycle-to-burn-height next-reward-cycle))
+            (next-signer-set-map (map-get? reward-cycle-signing-key-ids next-reward-cycle))
+        )
+
+        ;; Assert not already set by checking for 'none' map entry
+        (asserts! (is-none next-signer-set-map) (err ERR_NEXT_SIGNER_SET_ALREADY_REGISTERED))
+
+        ;; Assert in prepare phase (in the last 100 blocks of the current cycle)
+        (asserts! 
+            (and 
+                (>= block-height (- next-reward-cycle-burn-height u100))
+                (< block-height next-reward-cycle-burn-height))
+                (err ERR_NOT_IN_PREPARE_PHASE))
+
+        ;; Updates new signer-set
+        (var-set current-signer-set new-signer-set)
+    )
+)
+
+;; This function allows a node to register a list of key-ids to the current-signer-set
+;; This work is done one-at-a-time by the node & could require up to 4096 individual writes
+(define-private (update-current-signer (signer (buff 33)) (key-ids (list 4096 uint)))
+    (let 
+        (
+            (current-signers (var-get current-signer-set))
+            (next-reward-cycle (+ u1 current-pox-reward-cycle))
+        )
+
+        ;; Assert that signer exists in current list
+        (asserts! (is-some (index-of current-signers signer)) (err ERR_NOT_IN_SIGNER_SET))
+
+        ;; Assert that signer key-ids haven't been set yet (map entry is empty)
+        (asserts! (is-none (map-get? reward-cycle-signer-key-ids {cycle: next-reward-cycle, signer: signer})) (err ERR_KEY_IDS_ALREADY_SET))
+
+        ;; Q - do we need another check here that we're in the 100-block prepare phase?
```

Comment:
Prob not a bad idea to have this check. The node is the only thing that would ever call this function, but in theory we would never be setting the signer set except in the prepare phase.